src/transformers/models/ibert/utils/quant_modules.py

```diff
@@ -0,0 +1,538 @@
+        scaling_factor = 1 / 2 ** self.output_bit
+        return exp_int * scaling_factor, scaling_factor
+
+
+class IntLayerNorm(Module):
+    """
+    Class to quantize given LayerNorm layer
+
+    Parameters:
+    ----------
+    output_bit : int
+        Bitwidth for the LayerNorm output.
+    quant_mode : bool, default False
+        The mode for quantization. True for quantization.
+    force_dequant : str, default 'none'
+        Force dequantize LayerNorm if either 'layernorm' or 'nonlinear' is given.
+    """
+    def __init__(self,
+                 output_bit,
+                 quant_mode=False,
+                 force_dequant='none'):
+        super(IntLayerNorm, self).__init__()
+        self.quant_mode = quant_mode
+        if force_dequant in ['nonlinear', 'layernorm']:
+            logger.info("Force dequantize layernorm")
+            self.quant_mode = False
+
+        self.register_buffer('shift', torch.zeros(1))
+        self.output_bit = output_bit
+        self.max_bit = 32
+        self.dim_sqrt = None
+
+        self.activation = QuantAct(self.output_bit, quant_mode=self.quant_mode)
+        if not self.quant_mode:
+            pass
+        else:
+            self.weight_function = SymmetricQuantFunction.apply
+
+    def set_param(self, ln):
+        self.normalized_shape = ln.normalized_shape
+        self.eps = ln.eps
+        self.weight = Parameter(ln.weight.data.clone())
+        self.bias = Parameter(ln.bias.data.clone())
+
+    def set_shift(self, y_int):
+        with torch.no_grad():
+            y_sq_int = y_int ** 2
+            var_int = torch.sum(y_sq_int, axis=2, keepdim=True)
+            shift = (torch.log2(torch.sqrt(var_int / 2 ** self.max_bit)).ceil()).max()
+            shift_old = self.shift
+            self.shift = torch.max(self.shift, shift)
+            logger.info("Dynamic shift adjustment: {} -> {}".format(
+                int(shift_old), int(self.shift))
+            )
+
+    def overflow_fallback(self, y_int):
+        """
+        This fallback function is called when overflow is detected during training time,
+        and adjusts the `self.shift` to avoid overflow in the subsequent runs.
+        """
+        self.set_shift(y_int) # adjusts `self.shift`
+        y_int_shifted = floor_ste.apply(y_int / 2 ** self.shift)
+        y_sq_int = y_int_shifted ** 2
+        var_int = torch.sum(y_sq_int, axis=2, keepdim=True)
+        return var_int
+
+    def forward(self, x, scaling_factor=None):
+        if not self.quant_mode:
+            mean = x.mean(axis=2, keepdim=True)
+            y = x - mean
+            var = torch.mean(y ** 2, axis=2, keepdim=True)
+            x = y / torch.sqrt(self.eps + var)
+            x = x * self.weight + self.bias
+            return x, None
+
+        # compute sqrt of the feature dimension if it is the first run
+        if self.dim_sqrt is None:
+            n = torch.tensor(x.shape[2], dtype=torch.float) 
+            self.dim_sqrt = torch.sqrt(n).cuda()
```

Comment:
Does this only work on GPU? -> Is is not possible to run those functions on CPU?

Reply:
All calls to cuda should be avoided, we should use `to(x.device)` instead.